selfhost/codegen.jakt

```diff
@@ -36,49 +36,44 @@ fn is_return_allowed(anon allowed_control_exits: AllowedControlExits) -> bool =>
 
 struct ControlFlowState {
     allowed_exits: AllowedControlExits
-    /// Whether `break` and `continue` should use `return ExplicitValueOrControlFlow` instead of
-    /// C++'s `break`/`continue`.
-    passes_through_match: bool
     passes_through_try: bool
-    match_nest_level: usize
+    // whether the current scope is a match IIFE
+    directly_inside_match: bool = false
+    // whether any outer scope is a match IIFE
+    indirectly_inside_match: bool = false
 
     fn no_control_flow() -> ControlFlowState {
         return ControlFlowState(
             allowed_exits: AllowedControlExits::Nothing
-            passes_through_match: false
             passes_through_try: false
-            match_nest_level: 0
+            directly_inside_match: false
+            indirectly_inside_match: false
         )
     }
     fn enter_function(this) -> ControlFlowState {
         return ControlFlowState(
             allowed_exits: AllowedControlExits::JustReturn
-            passes_through_match: false
             passes_through_try: false
-            match_nest_level: .match_nest_level
+            indirectly_inside_match: .indirectly_inside_match
+            directly_inside_match: false
```

Comment:
That looks like a bug waiting to happen. Entering a function means the scope for return changes, which means that it should set "indirectly inside match" to false as well.